### PR TITLE
PlugIn rule added

### PR DESCRIPTION
**Pull request by andrepugni**
Hey Carlos,

I have added the PlugIn rule in the xaiUtils and a few tests for it. I hope everything is ok code-wise.

### CODE REVIEW

**Comment by cmougan** [comment]
could you add to the gitiginore the .idea?
Also the inspection profiles

They are local files that we don't want to have in the repo

**Comment by cmougan** [comment]
Also run a code style checker (black)
# for all files
python -m black * 
# for a specific one
python -m black file.py

File: test/plugin_test.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+X, y = make_blobs(n_samples=2000, centers=2, n_features=5, random_state=0)
+X_tr, X_te, y_tr, y_te = train_test_split(X, y, test_size=0.5, random_state=0)
+X_ood, y_ood = make_blobs(n_samples=1000, centers=1, n_features=5, random_state=0)
+
+
+def test_doc_examples():
+    """
+    Test it works
+    """
+    from sklearn.model_selection import train_test_split
```
Comment (cmougan) [inline test/plugin_test.py:16]: Double imports.

File: tools/xaiUtils.py [hunk 2]
```diff
@@ -258,3 +257,60 @@ def get_linear_coefs(self):
...
+    >>>     acc = accuracy_score(y_te[selected], preds[selected])
+    >>>     print("target coverage is: {}".format(1-clf.quantiles[level]))
+    >>>     print("coverage is: {}".format(coverage))
+    >>>     print("selective accuracy is: {}".format(acc))
+    """
+    def __init__(self, model, quantiles: list = [.01, .05, .10, .15, .20, .25], seed: int = 42):
+        self.quantiles = quantiles
+        self.seed = seed
+        self.thetas = None
+        self.model = copy.deepcopy(model)
```
Comment (cmougan) [inline tools/xaiUtils.py:304]: Why a deepcopy?
why not just self.model = model?

File: tools/xaiUtils.py [hunk 3]
```diff
@@ -258,3 +257,60 @@ def get_linear_coefs(self):
...
+
+    def fit(self, X, y, sample_weight=None):
+        self.classes_ = list(np.unique(y))
```
Comment (cmougan) [inline tools/xaiUtils.py:307]: Why do you do this?

File: tools/xaiUtils.py [hunk 4]
```diff
@@ -258,3 +257,60 @@ def get_linear_coefs(self):
...
+    >>> X, y = make_blobs(n_samples=2000, centers=2, n_features=5, random_state=0)
+    >>> X_tr, X_te, y_tr, y_te = train_test_split(X, y, test_size=0.5, random_state=0)
+    >>> X_ood,y_ood = make_blobs(n_samples=1000, centers=1, n_features=5, random_state=0)
+
+    >>> clf = PlugInRule(model=XGBRegressor())
+    >>> clf.fit(X_tr, y_tr)
+    >>> scores = clf.predict_proba(X_te)[:,1]
+    >>> preds = clf.predict(X_te)
+    >>> bands = clf.qband(X_te)
+    >>> for level in range(len(clf.quantiles)+1):
```
Comment (cmougan) [inline tools/xaiUtils.py:286]: Why this is in the doc code? Shouldnt this code be the minimum viable code?

Comment (cmougan) [reply tools/xaiUtils.py:286]: Why is not level a hyperparameter that you fit to the PlugIn?

File: test/plugin_test.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    """
+    Check that thetas are estimated after fit
+    """
+    X, y = make_blobs(n_samples=2000, centers=2, n_features=5, random_state=0)
+    X_tr, X_te, y_tr, y_te = train_test_split(X, y, test_size=0.5, random_state=0)
+    # X_ood,y_ood = make_blobs(n_samples=1000, centers=1, n_features=5, random_state=0)
+
+    clf = PlugInRule(model=XGBRegressor())
+    clf.fit(X_tr, y_tr)
+    assert clf.thetas is not None
```
Comment (cmougan) [inline test/plugin_test.py:65]: You are checking here that is not None.
What is the desired outcome?
I think theta can not be ever none, so this wont show an error.

File: test/plugin_test.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        selected = bands >= level
+        coverage = len(y[selected])/len(y)
+        acc = accuracy_score(y_te[selected], preds[selected])
+        print("target coverage is: {}".format(1-clf.quantiles[level]))
+        print("coverage is: {}".format(coverage))
+        print("selective accuracy is: {}".format(acc))
+
+def test_plugin_fitted():
+    """
+    Check that no NaNs are present in the shap values.
```
Comment (cmougan) [inline test/plugin_test.py:42]: This doc does not represent what is done

File: test/plugin_test.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+from sklearn.model_selection import train_test_split
+from sklearn.datasets import make_blobs
+import numpy as np
+
+X, y = make_blobs(n_samples=2000, centers=2, n_features=5, random_state=0)
+X_tr, X_te, y_tr, y_te = train_test_split(X, y, test_size=0.5, random_state=0)
+X_ood, y_ood = make_blobs(n_samples=1000, centers=1, n_features=5, random_state=0)
+
+
+def test_doc_examples():
```
Comment (cmougan) [inline test/plugin_test.py:12]: There is no assert, you are just check-ing that this runs, you don't expect nothing to happen

**Comment by cmougan** [comment]
inspection files too in the gitignore

**Comment by andrepugni** [comment]
Shold have done it

**Comment by cmougan** [comment]
I will come back to your office at 15.30 and we have a look :)

**Review by cmougan — APPROVED** [review]

**Pull request merged**